rust/src/sdk/api.rs

```diff
@@ -87,85 +84,174 @@ macro_rules! PrimModel {
     }
-    let client = if let Ok(client) = Client::new(period as u64) {
+    let runtime = if let Ok(runtime) = Builder::new()
+        .threaded_scheduler()
+        .core_threads(min_threads as usize)
+        .max_threads(max_threads as usize)
+        .thread_name("xain-fl-client-runtime-worker")
+        .enable_all()
+        .build()
+    {
+        runtime
+    } else {
+        return ptr::null_mut() as *mut FFIClient;
+    };
+    let client = if let Ok(client) = runtime.enter(move || Client::new(period)) {
         client
     } else {
-        // TODO: add error handling
-        panic!("participant initialization failed")
+        return ptr::null_mut() as *mut FFIClient;
     };
+    Box::into_raw(Box::new(FFIClient { runtime, client }))
+}
 
-    // TODO: actually start the client, requires tokio running
-
-    #[repr(C)]
-    struct Input {
-        _round_started: bool,
-        _participant_initialized: bool,
-        _model_cached: bool,
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Starts the [`Client`].
+///
+/// The [`Client`]'s tasks are executed in an asynchronous [`Runtime`].
+///
+/// Takes a `callback(state, code)` function pointer and a void pointer to the `state` of the
+/// callback. The callback will be triggered when the client must be stopped because of a panic or
+/// error. See the [errors] section for the error `code` definitions.
+///
+/// # Errors
+/// Ignores null pointer `client`s and triggers the callback immediately. Triggers the callback
+/// with one of the following error codes in case of a [`Client`] panic or error:
+/// - `-1`: client didn't start due to null pointer
+/// - `0`: no error
+/// - `1`: client panicked due to unexpected/unhandled error
+/// - `2`: client stopped due to error [`ParticipantInitErr`]
+/// - `3`: client stopped due to error [`ParticipantErr`]
+/// - `4`: client stopped due to error [`DeserialiseErr`]
+/// - `5`: client stopped due to error [`GeneralErr`]
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+///
+/// If the callback is triggered because of a panicking client, it is the users responsibility
+/// to not access possibly invalid state and to drop the client. If certain parts of the state
+/// can be guaranteed to be valid, they may be read before dropping.
+///
+/// [errors]: #errors
+/// [`ParticipantInitErr`]: ../../client/enum.ClientError.html#variant.ParticipantInitErr
+/// [`ParticipantErr`]: ../../client/enum.ClientError.html#variant.ParticipantErr
+/// [`DeserialiseErr`]: ../../client/enum.ClientError.html#variant.DeserialiseErr
+/// [`GeneralErr`]: ../../client/enum.ClientError.html#variant.GeneralErr
+pub unsafe extern "C" fn run_client(
+    client: *mut FFIClient,
+    callback: unsafe extern "C" fn(*mut c_void, c_int),
+    state: *mut c_void,
+) {
+    if client.is_null() {
```

Comment:
if `client` is `null` there clearly is a bug in the client code, and I'm not sure we should try to call the callback in that case.

Reply:
i'm not sure myself to be honest. the error code is part of the callback, so the user could match the code and act accordingly. otherwise we might just return the error code from `run_client()` and call the callback without the error code, but the error code might be useful for callbacks for regular errors. or we could just have a `!client.is_null()` in the beginning and ignore null pointers at all. what do you think?